src/value.rs

```diff
@@ -90,6 +101,88 @@ impl AllocatedValue {
     }
 }
 
+impl SignedInteger {
+    pub fn zero() -> Self {
+        SignedInteger::Positive(0)
+    }
+
+    pub fn to_u64(&self) -> Option<u64> {
+        match self {
+            SignedInteger::Positive(x) => Some(*x),
+            SignedInteger::Negative(_) => None,
+        }
+    }
+
+    pub fn sign(&self) -> u8 {
+        match self {
+            SignedInteger::Positive(_) => 1,
+            SignedInteger::Negative(_) => 0,
+        }
+    }
+
+    fn to_i128(&self) -> i128 {
+        match self {
+            SignedInteger::Positive(x) => (*x).into(),
+            SignedInteger::Negative(x) => -1 * i128::from(*x),
+        }
+    }
+}
+
+impl From<u64> for SignedInteger {
+    fn from(u: u64) -> SignedInteger {
+        SignedInteger::Positive(u)
+    }
+}
+
+impl Into<Scalar> for SignedInteger {
+    fn into(self) -> Scalar {
+        match self {
+            SignedInteger::Positive(x) => x.into(),
+            SignedInteger::Negative(x) => Scalar::zero() - Scalar::from(x),
+        }
+    }
+}
+
+impl Add for SignedInteger {
+    type Output = SignedInteger;
+
+    fn add(self, rhs: SignedInteger) -> SignedInteger {
+        match (self, rhs) {
+            (SignedInteger::Positive(l), SignedInteger::Positive(r)) => {
```

Comment:
Would it be simpler if we convert to i128, do addition and then convert back to SignedInteger? We can then factor out this into a private from_i128 method.

Reply:
But i'm impressed by the precision of this method.